### ISSUE TITLE
all-project search is very slow when large number of repositories are indexed

### ISSUE THREAD

**Issue by rahul-chhabra** [OpenGrok/OpenGrok#1116]
Hi,
The search is very slow when large number of repositories are indexed.
Every search is taking around 15000-25000 milliseconds.

These are some details:
-----------------------------------------------------------------------------------------------
Number of Repos: ~700
Repos Size:		11G
Index Size:		2.5G
Time taking to search: 15000-20000 milliseconds every search
Time taken for indexing: 10 minutes
Opengrok version:	opengrok-0.12.1.5
Tomcat version: apache-tomcat-7.0.69
apache-tomcat-7.0.69/conf/server.xml:	maxHttpHeaderSize="65536"

Could you please tell if there is any configuration I can change in opengrok or apache tomcat to make search faster ? 
Previously it used to take 100-300 milliseconds only when number of repositories were around 50.

Thanks,
Rahul

**Comment by tarzanek** [comment]
and you search over ALL repos at the same time?
if yes, then this might be expected and needs to be improved
if you search just through one repo, then this slow down isn't really expected ... (unless we have a leak in tomcat ... )

**Comment by rahul-chhabra** [comment]
Thanks @tarzanek for quick reply. 
Yes, when ALL repositories had selected I am facing this issue.
On single repository Opengrok is able to return result in average 250 milliseconds.

1. Is there any workaround by which I can improve this performance ?
Even the slightest improvement would be helpful.

2. Is this slowness solely because of large number of repositories (~700) in GET request ?

**Comment by tarzanek** [comment]
I fear it is because of that, top I've seen is 100 projects in one instance.
Could you divide your projects to more opengroks?
Or they are branches from one repository?
Anyways you might try 0.13 rc, but it will only give you small boost.
I guess we need some improvements when large count of projects is used ...

**Comment by vladak** [comment]
It would be useful to know what limits the scaling in this case. Is it I/O or CPU ? How many CPUs are in the system ? How many hardware threads/execution units do the CPUs in the system have ? Basic tools in given system can help to find out (dtrace, iostat, etc.).

**Comment by rahul-chhabra** [comment]
Thanks @tarzanek I will try your option too, dividing into multiple opengroks and see how it goes. 
However I will still really like to have ability to search ALL repos (_**in a superfast way**_) within a single opengrok.

Hi @vladak 
It is pretty free Linux machine and no other process currently running  on it.
I have got stats using **iostat.**
There are 4 CPU's and around total 4G of memory. 
I am not Linux expert but if you can tell me exact command with arguments, i can let you know the output.

Many thanks.

**Comment by tarzanek** [comment]
4g of mem is too low. 8g would be better. 16g ideal ...
Jvm grows to 4g easily, ctags can grow to 1-2g too ...

**Comment by tarzanek** [comment]
That is for indexing. Since you serve requests from tomcat - give it more heap too so 64bit server jvm and at least 4g mem ... so in your case 16g of ram might be minimum ...

**Comment by tarzanek** [comment]
Also tomcat8 + jdk8 might help.
But memory is the base ... so without more mem anything else is just minimal impact.

**Comment by rahul-chhabra** [comment]
thanks @tarzanek 
I will get memory increase to 16G and allocate more heap space to tomcat. 
Maybe you can close question till then, I will reopen once I have results with these changes, which might take a week.

**Comment by vladak** [comment]
To solve this problem, we need to answer multiple questions: is it scaling problem in OpenGrok or Lucene ? Is the bottleneck in the operating system or hardware ?

The good thing in the observed performance is that the high search times (tens of seconds) give you the ability to see what is going on in the system quite easily. I see that @tarzanek already suggested to use the "scientific" methodology (number 4 in http://www.brendangregg.com/methodology.html) with the hypothesis that the bottleneck is system memory. If that proves insufficient, the USE method (http://www.brendangregg.com/usemethod.html) can be used to detect all problems in the system. Since you are running Linux, the table on http://www.brendangregg.com/USEmethod/use-linux.html should help you to see what tools can be used to identify the problematic resources. In your case the most interesting facts will probably be swapping and I/O activity and CPU utilization/saturation. This should help us understand what is the system really doing during these 15-20 seconds when it is searching.

As for the CPU load, each multi-project search uses `2 * availableProcessors()` threads (I found issue with sub-optimal use of the threads and fixed it in #1119) to perform the search on all indexes in parallel. Assuming, you're running on some recent x86 box, those 4 CPUs would probably be 2 execution units with HypetThreading. It is a question whether HT really helps for such load however assuming there is not a problem to get these 2.5 GB into memory the search should be really fast.

As for the memory, it depends on what method will be used to load the indexes into memory. OpenGrok lets Lucene to select the method automatically. This is visible in SearchEngine.java#searchMultiDatabase which is really the core of the search:

```
207        for (Project project : root) {
208            IndexReader ireader = (DirectoryReader.open(FSDirectory.open(new File(droot, project.getPath()).toPath())));
209            subreaders[ii++] = ireader;
210        }
211        MultiReader searchables = new MultiReader(subreaders, true);
212        if (Runtime.getRuntime().availableProcessors() > 1) {
213            searcher = new IndexSearcher(searchables,
214                RuntimeEnvironment.getInstance().getSearchExecutor());
215        } else {
216            searcher = new IndexSearcher(searchables);
217        }
218        collector = TopScoreDocCollector.create(hitsPerPage * cachePages);
219        searcher.search(query, collector);
```

`FSDirectory.open` will select the best method as documented here: https://lucene.apache.org/core/6_1_0/core/org/apache/lucene/store/FSDirectory.html. So far the best method for accessing large indexes is the MMapDirectory implementation which uses mmap() to map the whole index into memory. Some good explanation can be found on this blog: http://blog.thetaphi.de/2012/07/use-lucenes-mmapdirectory-on-64bit.html To make sure this method is used, one can add `-d64` switch to Tomcat to make it run as 64-bit process.

The above code is called through few simple layers from the webapp so any OpenGrok involvement is minimal I'd say. If there is a problem, it is either in Lucene, operating system or the hardware.

When I boot a recent version of Solaris in a kernel zone, it has cca 2.5 GB of memory free. This should be enough to hold index of your size. Some memory will be of course needed for Java/OpenGrok. If multiple threads start random I/O in multiple distinct mmap-ed files, they will be competing for file system/page cache however given the amount of free memory available, I would not expect any serious performance degradation compared to just a few repositories. If you can share the nature of the repositories (number of files, distribution of file sizes, language) and all-project search in more detail (are you performing full text/definition/symbol/file path search and how complicated are the search terms), I can give it a try to reproduce similar scenario in house.

**Comment by rahul-chhabra** [comment]
Thanks @vladak  for detail explanation.
I have requested and will get memory increase in my linux server in couple of days.
After that, I will increase memory allocation and will share the results and other mentioned details soon.

**Comment by vladak** [comment]
I tried to replicate the setup in house: Solaris 11.3 x86 running in kernel zone, has 4 GB of RAM, 4 virtual CPUs running at 3 GHz. Used OpenGrok 0.13-rc1. The host in which the zone is running is SUN FIRE X4170 M2 SERVER with 2 x Xeon X5675 @ 3.07GHz. The storage on the machine is based on raidz1 volume with 7 disks (note that raidz1 is not the performance winner). The data (both sources and index) inside the zone is also stored on ZFS dataset.

I cloned the openssh-portable repo from github and multiplied it with some variance using:

```
i=0; while [ $i -le 5 ]; do git tag | while read tag; do 
    git clone /var/tmp/openssh-portable /var/opengrok/src/ossh-$tag-$i; 
    cd /var/opengrok/src/ossh-$tag; git reset --hard $tag; done; i=$( expr $i + 1 ); 
done
```

This created some 774 repositories. The index has total 2.4 GB.

Using single word in full text search, the times were from 10 to 20 seconds, sometimes even higher numbers (80 seconds). I can see that the filesystem cache has grown to the max and free memory is at its minimum:

```
Page Summary                 Pages             Bytes  %Tot
----------------- ----------------  ----------------  ----
Kernel                      430958              1.6G   41%
ZFS Metadata                 98970            386.6M    9%
ZFS File Data               230894            901.9M   22%
Anon                        167416            653.9M   16%
Exec and libs                  873              3.4M    0%
Page cache                    8010             31.2M    1%
Free (cachelist)              7391             28.8M    1%
Free (freelist)              29349            114.6M    3%
Total                      1047965              3.9G
```

No swapping activity, though.

When I perform the search (usually for some well-known keyword in full text search like `foo`, `function`, `definition` etc. and never using the same) it is obvious that the workload is I/O bound: I can see busy times from `iostat` to be around 100% and confirmed with `truss` that the webapp is indeed mmap-ing index files into memory.

The pattern is nicely visible in `vmstat` output. This is a 1 second progression when performing one of these long searches:

```
 kthr      memory            page            disk          faults      cpu
 r b w   swap  free  re  mf pi po fr de sr zb -- -- --   in   sy   cs us sy id
 0 0 0 1110780 257316 19 241 0  0  0  0  0 213 0  0  0 1091 1524 1065  0  2 97
 0 0 0 1110780 255672 14 206 0  0  0  0  0 186 0  0  0  980 1330  915  0  2 97
 0 0 0 1104628 247908 15 132 0  0  0  0  0 110 0  0  0  770  854  640  0  1 98
 1 0 0 1098480 240404 7 194  0  0  0  0  0 185 0  0  0 1001 1330  945  0  2 97
 1 0 0 1092204 232512 11 125 0  0  0  0  0 112 0  0  0  741  824  632  0  2 98
 0 0 0 1085864 225240 23 228 0  0  0  0  0 207 0  0  0 1073 1448 1041  0  1 98
 0 0 0 1077668 215080 24 227 0  0  0  0  0 204 0  0  0 1062 1445 1026  0  2 98
 1 0 0 1067268 202936 16 131 0  0  0  0  0 116 0  0  0  772  872  661  0  1 98
 0 0 0 1061120 195628 12 97  0  0  0  0  0 88  0  0  0  673  705  553  0  1 98
 0 0 0 1059072 193056 5  68  0  0  0  0  0 56  0  0  0  537  461  392  0  1 99
 0 0 0 1056952 190556 8 171  0  0  0  0  0 153 0  0  0  892 1104  812  0  1 98
 1 0 0 1050776 182248 24 224 0  0  0  0  0 188 0  0  0  996 1435  981  0  2 97
 0 0 0 1042580 172188 9 179  0  0  0  0  0 161 0  0  0  842 1180  780  0  1 98
 0 0 0 1036188 164324 16 239 0  0  0  0  0 211 0  0  0 1086 1580 1093  0  3 96
 0 0 0 1027996 154492 18 230 0  0  0  0  0 165 0  0  0  938 1223  908  0  2 97
 0 0 0 1021852 146652 22 272 0  0  0  0  0 232 0  0  0 1164 1668 1172  0  3 96
 1 0 0 1011480 134192 9 193  0  0  0  0  0 170 0  0  0  977 1206  953  0  3 96
 0 0 0 1003288 124264 15 188 0  0  0  0  0 174 0  0  0  965 1226  941  0  2 96
 1 0 0 992876 136288  1 519  0  0  0  0  0 157 0  0  0 1497 3081  856  1  4 95
```

We can see that the CPUs are mostly idle but the disk is chugging along. There are some minor faults (need to establish mappings for pages in memory) and some page reclaims but nothing dramatic.

Here's captured end of the search in `iostat`, again with 1 second period:

```
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   62.0    8.0 7681.9   87.0  0.0  8.6  123.4   0 100 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   58.0    0.0 7423.2    0.0  0.0  9.0  155.1   0 100 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   44.0    0.0 5631.9    0.0  0.0  9.0  203.6   0 100 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   58.0    2.0 7424.2    3.0  0.0  9.3  155.0   0 100 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   49.0   68.0 5537.7  178.0  0.0  6.8   58.3   0  81 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0    9.0    0.0  898.0    0.0  0.0  0.2   16.7   0  15 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   90.0    0.0 3758.5    0.0  0.0  1.0   11.2   0  97 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0   81.3    0.0  207.1    0.0  0.0  1.0   11.8   0  96 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0    5.0    0.0  390.3    0.0  0.0  0.0    4.9   0   2 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0    0.0    8.0    0.0  335.5  0.0  0.3   43.0   0   9 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0    5.0  169.0   18.0 8988.0  0.0  3.1   17.7   0  97 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0    3.0    6.0   17.0    4.0  0.0  0.1   11.1   0   6 
                 extended device statistics                 
device    r/s    w/s   kr/s   kw/s wait actv  svc_t  %w  %b 
zvblk0    0.0    0.0    0.0    0.0  0.0  0.0    0.0   0   0 
```

The busy percentage is around 100%, sometimes there is even some saturation (queue wait times in the %w column).

**Comment by vladak** [comment]
After increasing the memory of the zone to 8 GB the times stabilized under 5 seconds after couple of queries. 

Shortly after the reboot memory usage looked like this:

```
Page Summary                 Pages             Bytes  %Tot
----------------- ----------------  ----------------  ----
Kernel                      236578            924.1M   11%
ZFS Metadata                 48344            188.8M    2%
ZFS File Data               102192            399.1M    5%
Anon                        103004            402.3M    5%
Exec and libs                 1675              6.5M    0%
Page cache                   15893             62.0M    1%
Free (cachelist)             11894             46.4M    1%
Free (freelist)            1559561              5.9G   74%
Total                      2096541              7.9G
```

After performing couple of search requests I can see that ZFS cache has grown however there is still some free memory:

```
Page Summary                 Pages             Bytes  %Tot
----------------- ----------------  ----------------  ----
Kernel                      254028            992.2M   12%
ZFS Metadata                 70717            276.2M    3%
ZFS File Data               355412              1.3G   17%
Anon                        175712            686.3M    8%
Exec and libs                 1631              6.3M    0%
Page cache                   15572             60.8M    1%
Free (cachelist)             87640            342.3M    4%
Free (freelist)            1120102              4.2G   53%
Total                      2096541              7.9G
```

and I can see that new queries still generate significant I/O activity so the ZFS ARC still has some growing to do.

To summarize, adding memory definitely helps after file system caches get warmed up. For ultimate performance, I'd place the index on SSD if the cost is not so big issue.

**Comment by tarzanek** [comment]
@vladak nice sleuthing and nice reading :)
Disk i/o is huge issue thats'why we tried to limit the traversal and reread of same file with @mrate and his scope fixes.
When you do index on your setup - can we count how many times one file is accessed? When indexing from scratch of course ...

**Comment by vladak** [comment]
@tarzanek let's take this offline. The workloads of the indexer and searcher are different and I would like to avoid confusing this issue even more.

**Comment by rahul-chhabra** [comment]
Hi @vladak and @tarzanek,
I have got memory increase to 16G and included the below parameters in apache tomcat setenv.sh 
`export JAVA_OPTS="-d64 -Xms1024m -Xmx8192m -server"`

After couple of searches, the all-project search time has significantly come down to around 7,000 milliseconds from earlier 20,000 milliseconds.

Is there any other parameter that I can change, which will improve performance further ?

**Comment by vladak** [comment]
Good to see that the hypothesis was proved correct. 

As for the tuning, this is getting tricky. The ultimate cure would be to place the index wholly into RAM. The previously referenced blog post advises against using RAMDirectory in Lucene as it places the index data into Java heap might end up being swapped to disk (if the file system cache is prioritized in given operating system) and that would ruin the whole point. Anyhow, this would require changing OpenGrok code to use specific object instead of FSDirectory.

Another alternative would be to create a ramdisk (assuming it will never get swapped out) a place the data directory contents there. This has the downside that it would have to be synced after each reindex. Some scripting should take of that, though.

http://blog.mikemccandless.com/2010/07/lucenes-ram-usage-for-searching.html mentions how Lucene 4 (0.13 uses 6.x BTW) improved memory footprint drastically for data which need to be resident in RAM and advises to turn off swapping for Linux so the fight between swapping and file system cache is minimized so you might want to try that.

Lastly, there might be some ways to tune MMAPDirectory futher. Lot's of systems have private mmap() syscall flags which might help. That would be a subject of some research for sure.

Also, given that the index files do not seem to be fully loaded into RAM even after couple of warm-up searches, one could try some desperate measures like writing a small program which would mmap() all the index files into RAM and read them byte by byte in hope that file system caches will be populated with the data. The file system itself might also do some optimization, like read-ahead if it notices sequential pattern (ZFS does that).

**Comment by vladak** [comment]
http://serverfault.com/questions/43383/caching-preloading-files-on-linux-into-ram has some pieces of advice on how can the filesystem cache be populated, in particular it contains reference to the vmtouch tool which does exactly of what I proposed above - mmap()s all files in a directory and reads them in. This of course depends on how big the cache can grow. Tts documentation contains other tuning tips.

**Comment by vladak** [comment]
I did some experiments with `vmtouch`. Run it against the index like so:

```
# ./vmtouch -t /var/opengrok/data/index
           Files: 10062
     Directories: 775
   Touched Pages: 580164 (2G)
         Elapsed: 101.85 seconds
```

the ZFS data cache grew from:

```
Page Summary                 Pages             Bytes  %Tot
----------------- ----------------  ----------------  ----
Kernel                      267333              1.0G   13%
ZFS Metadata                 74386            290.5M    4%
ZFS File Data               418783              1.5G   20%
Anon                        160365            626.4M    8%
Exec and libs                 1602              6.2M    0%
Page cache                   14955             58.4M    1%
Free (cachelist)             94262            368.2M    4%
Free (freelist)            1048072              3.9G   50%
Total                      2096541              7.9G
```

by 1.5 GB to:

```
Page Summary                 Pages             Bytes  %Tot
----------------- ----------------  ----------------  ----
Kernel                      277376              1.0G   13%
ZFS Metadata                 74839            292.3M    4%
ZFS File Data               789785              3.0G   38%
Anon                        166424            650.0M    8%
Exec and libs                 1616              6.3M    0%
Page cache                   15979             62.4M    1%
Free (cachelist)            463371              1.7G   22%
Free (freelist)             290111              1.1G   14%
Total                      2096541              7.9G
```

which reduced all-project searches under 3 seconds.

The webapp is still doing some I/O - mostly the files which are displayed in search results which is okay and expected. I confirmed this with `rwsnoop` script from Dtrace Toolkit.

Using `dvmstat` script from Dtrace Toolkit we can see VM statistics for particular process, the webapp itself. Single all-projects search for previously unsearched term looks like this:

```
root@S11:/usr/dtrace/DTT# ./dvmstat -p `svcs -H -o fmri -p tomcat8 | tail -1 | awk '{ print $2 }'`
    re   maj    mf   fr  epi  epo  api  apo  fpi  fpo     sy
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     29
     0     0     0    0    0    0    0    0    0    0     26
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     26
 34148     0 34324    0    0    0    0    0    0    0  45742
 34440     0 35136    0    0    0    0    0    0    0  35083
   172     0   176    0    0    0    0    0    0    0   5607
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     25
     0     0     0    0    0    0    0    0    0    0     22
     0     0     0    0    0    0    0    0    0    0     26
     0     0     0    0    0    0    0    0    0    0     26
     0     0     0    0    0    0    0    0    0    0     25
```

The cca 2 second search is clearly visible there (`dvmstat` period is 1 second).

The fields are interpreted as follows:

```
# FIELDS: 
#               re      page reclaims           Kbytes
#               maj     major faults            Kbytes
#               mf      minor faults            Kbytes
#               fr      page frees              Kbytes
#               epi     executable page ins     Kbytes
#               epo     executable page out     Kbytes
#               api     anonymous page ins      Kbytes
#               apo     anonymous page outs     Kbytes
#               fpi     filesystem page ins     Kbytes
#               fpo     filesystem page outs    Kbytes
#               sy      system calls            number
```

so bunch of page reclaims, minor faults and syscalls. No page-ins were observed so the file system cache seems to be properly warmed.

Using `mpstat` it looks like CPU is not fully ulitized:

```
 CPU minf mjf xcal  intr ithr  csw icsw migr smtx  srw syscl  usr sys  st idl
   0    0   0    0   217    6   42    0    2    0    0    24    0   0   1  99
   1    0   0    0    69    3   50    0    4    0    0     3    0   0   0 100
   2    0   0    0    86    6   58    0    0    0    0     2    0   0   0 100
   3    0   0    0    48   20   50    1    3    0    0    36    0   0   0 100
 CPU minf mjf xcal  intr ithr  csw icsw migr smtx  srw syscl  usr sys  st idl
   0 1544   0    8   229    3   69    0    2    0    0  7920    4   4   0  92
   1    0   0    0    64    6   65    0    4    0    0    34    0   0   0 100
   2    4   0    0   153   44  114    0    3    0    0   118    1   1   1  97
   3    0   0    0   132   19  201    0    5    0    0   279    0   0   1  99
 CPU minf mjf xcal  intr ithr  csw icsw migr smtx  srw syscl  usr sys  st idl
   0 15805   0   48   229    5   39   10    6   13    0 71408   26  52   0  22
   1  193   0    0    78    9  123    3   15   13    0   130    3   1   0  96
   2  133   0    0   160   18  196    3   23   14    0   120    3   0   0  97
   3  228   0    0   111   15  148    1   21   12    0   151    2   1   0  97
 CPU minf mjf xcal  intr ithr  csw icsw migr smtx  srw syscl  usr sys  st idl
   0    1   0    0   245   11   83    0   11    0    0    43    2   0   0  98
   1   57   0    0    73    7   78    2    8    2    0   395    5   1   0  94
   2    0   0    0   156   46  122    5    7    2    0    79    2   1   0  97
   3    2   0    0    99    6  156    3   10    2    0   203    3   0   0  97
 CPU minf mjf xcal  intr ithr  csw icsw migr smtx  srw syscl  usr sys  st idl
   0    0   0    0   234    2   69    0    8    4    0    57    0   0   1  99
   1    0   0    6    61   15   73    6   11    2    0  4740    2   2   1  95
   2    0   0    0   149   43  125    0   10    4    0   109    0   0   1  99
   3    0   0    0    57    7   69    0    8    4    0    67    0   0   1  99
```

Now it would be good to drill down and see what is the process up to during these 3 seconds when it is performing search, i.e. do some profiling and produce flame graphs or such. This is beyond of what I can do right now (skill and time wise).

**Comment by vladak** [comment]
Trying to see what is going on in terms of threads, I captured 2 second view during search:

```
# TERM=console prstat -n 1000 -s cpu -L -m -p `svcs -H -o fmri -p tomcat8 | tail -1 | awk '{ print $2 }'` 1 | grep -v ^Total

...

   PID USERNAME USR SYS TRP TFL DFL LCK SLP LAT VCX ICX SCL SIG PROCESS/LWP   
   646 webservd  10  12 0.0 0.0 0.0  78 0.0 0.0   3   3 25K   0 java/1592
   646 webservd 0.9 0.0 0.0 0.0 0.0  99 0.0 0.1   7   1  11   0 java/11
   646 webservd 0.1 0.3 0.0 0.0 0.0 0.0  99 0.6  72   0 217   0 java/1575
   646 webservd 0.3 0.0 0.0 0.0 0.0 100 0.0 0.0   3   0   8   0 java/13
   646 webservd 0.2 0.0 0.0 0.0 0.0 100 0.0 0.1   7   0   9   0 java/12
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   5   0  18   0 java/1596
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.1   5   0  18   0 java/1594
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   4   0  16   0 java/1595
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   4   0  16   0 java/1597
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   4   0  16   0 java/1598
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   3   2  13   0 java/7
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0  20   0  20   0 java/15
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   3   0  12   0 java/1591
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   3   0  12   0 java/1580
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   3   0  12   0 java/1599
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   3   0  12   0 java/1593
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   9   0 java/1576
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   1   0 java/16
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1578
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1577
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1574
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/19
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/18
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1579
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1572
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1573
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/17
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/14
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/10
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/9
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/8
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/6
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/5
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/4
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/3
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/2
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1
   PID USERNAME USR SYS TRP TFL DFL LCK SLP LAT VCX ICX SCL SIG PROCESS/LWP   
   646 webservd  20  42 0.0 0.0 0.0  38 0.0 0.1   5   9 54K   0 java/1592
   646 webservd 2.3 0.0 0.0 0.0 0.0  98 0.0 0.0   5   2   9   0 java/4
   646 webservd 2.3 0.0 0.0 0.0 0.0  98 0.0 0.0   3   0   7   0 java/3
   646 webservd 2.2 0.0 0.0 0.0 0.0  98 0.0 0.0   4   1   7   0 java/5
   646 webservd 2.2 0.0 0.0 0.0 0.0  98 0.0 0.0   3   0   6   0 java/6
   646 webservd 0.4 0.2 0.0 0.0 0.0  99 0.0 0.1   8   5  33   0 java/1880
   646 webservd 0.4 0.1 0.0 0.0 0.0  99 0.0 0.1   8   2  30   0 java/1882
   646 webservd 0.3 0.1 0.0 0.0 0.0  99 0.0 0.6   5   2  27   0 java/1881
   646 webservd 0.3 0.1 0.0 0.0 0.0  99 0.0 0.5   8   0  30   0 java/1883
   646 webservd 0.3 0.1 0.0 0.0 0.0 100 0.0 0.0   5   0  24   0 java/1885
   646 webservd 0.2 0.1 0.0 0.0 0.0 100 0.0 0.0   4   1  11   0 java/7
   646 webservd 0.2 0.1 0.0 0.0 0.0 100 0.0 0.1   8   0  30   0 java/1884
   646 webservd 0.2 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   2   0 java/12
   646 webservd 0.1 0.0 0.0 0.0 0.0 100 0.0 0.0   5   0  27   0 java/1887
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.2   8   0  30   0 java/1886
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   8   0  27   0 java/1888
   646 webservd 0.0 0.0 0.0 0.0 0.0  99 0.0 0.6   5   0  27   0 java/1889
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.1  17   0  17   0 java/15
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   5   0  12   0 java/18
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   1   0 java/9
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   2   0 java/8
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1578
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1575
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1574
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   1   0 java/16
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/1577
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   1   0   1   0 java/19
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   2   0 java/11
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1599
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1593
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1595
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1591
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1597
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1598
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1580
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1579
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1576
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1596
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1572
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1573
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/1594
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/17
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/14
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   1   0   1   0 java/13
   646 webservd 0.0 0.0 0.0 0.0 0.0 100 0.0 0.0   0   0   0   0 java/10
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/2
   646 webservd 0.0 0.0 0.0 0.0 0.0 0.0 100 0.0   0   0   0   0 java/1
```

Thread 1592 which is most probably the webapp request handler is doing lots of syscalls and consumes most CPU while the other most active threads (probably the search threads) consume much less CPU, the rest of threads is mostly idle.

**Comment by vladak** [comment]
Looking with `truss` on what is the request handler thread doing, it is basically opening and mmap-ing all the index files. I wonder if the cycle on lines 207-210 in the `searchMultiDatabase` code above could be done only after reindex or webapp startup. This could save 10s of thousands of syscalls in this case.

**Comment by vladak** [comment]
As for MmapDirectory tuning, Lucene has a method `setPreload(boolean)`. If set to true, it will call MappedByteBuffers's load method (https://lucene.apache.org/core/5_5_0/core/org/apache/lucene/store/MMapDirectory.html#setPreload(boolean)). If I understand it correctly, it will do similar thing as `vmtouch`. Given this might lead to heavy disk I/O I would be reluctant to add it to OpenGrok code base, even as an option.

**Comment by rahul-chhabra** [comment]
Hi @vladak 
I have tried to use vmtouch, but it has limitation it can load upto 1 GB of index in memory and after that need to change ulimit, which needs root permission. I will try further on that.

There are couple of things:
1. Could you please let know, in which file this lucene option- setPreload can be changed in opengrok codebase, I will try to build locally and see how it works for my usecase. 

2. Since all-project search is slow and it takes significant time, is it possible to create an enhancement where end-user gets some visual indication that search is in-progress ?
Also, when we do consecutive searches, may be we can clear results of first search and show some in-progress icon until results are displayed.
Also, when user directly tries opengrok link with all-project search, this will be very useful.

Thanks.

**Comment by vladak** [comment]
Hi @rahul-chhabra , I wonder what limit does `vmtouch` bump into ? Possibly max virtual memory or max memory size ? In that case, loading the files into memory from the webapp might hit the same limit.

Anyhow, the `setPreload()` method belongs to the MmapDirectory object so I'd place it into `SearchEngine.java#searchMultiDatabase()` after that object is instantiated. You would have to change `FSDirectory` to `MmapDirectory` as it extends the object definition.

As for displaying the progress indication (turning wheel, jumping animal, etc.) during search I am a bit torn. OpenGrok is touting its speed so one could argue that any deployment which is slow is under provisioned. Anyhow, feel free to submit new issue for that and it can be discussed further. Pull requests are welcome too :-)

**Comment by rahul-chhabra** [comment]
hi @vladak 
> I wonder what limit does vmtouch bump into ? Possibly max virtual memory or max memory size ? In that case, loading the files into memory from the webapp might hit the same limit.

Regarding this, vmtouch has limitation that it can load upto 1 GB of index in memory and after that need to change ulimit, which needs root permission. More details about this here: [https://github.com/hoytech/vmtouch/issues/18] 18 and [https://github.com/hoytech/vmtouch/issues/35] 35. 
In my system, there is 16 GB memory, 8 GB allocate to Tomcat and 8 GB is almost free, since no other process is running.

**Comment by vladak** [comment]
The first issue is about constraints for locked memory. According to the discussion the only cure for that is to raise the limit (the issue is closed after all). The second issue does not have a clear root cause. There is just some speculation about vmtouch bumping against max number of open files which can be caused by not close()ing the files after they are mmap()'ed which can be easily fixed in vmtouch source code. It is not clear to me why you cannot simply raise the limits in your setup. Anyhow, with the results being hidden now in `master` after @tulinkry 's PR and with the changes for reuse of IndexSearcher objects on the way, I feel we have reached the point of debating this fully, hence closing. Both changes will be in 0.13-rc3 once it is released.
